add minId overloads

kijanawoodard opened the pull request:
Fixes #1718

Initial commit. Still orienting on how to contribute.
It's building. Tests pass. 

Before I continue, does the diff look sane?

I think I got the "public api" situated correctly.

The strangest thing I ran into was this error about default parameters. I'm not sure if this is pushing me towards adding a new method rather than an overload, e.g. StreamTrimWithMinId. I took out the default CommandFlags to build successfully.

`error RS0026: Symbol 'StreamTrim' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'`

I'm having trouble running the tests locally. I tried running the tests before making any changes and ~2/3 passed. It might be because I don't have all the versions of .net installed.

kijanawoodard commented:
I can't tell what's wrong in the appveyor build. It says zero tests failed, but still errors.

kijanawoodard commented:
I was able to get devcontainer working on WSL enough to build, but the tests can't connect to redis in the devcontainer.

I'll post my changes in case anyone else can follow the breadcrumbs.

I needed to enable Ubuntu integration in docker desktop
![image](https://github.com/user-attachments/assets/9af75acb-f242-4729-9b16-159efa584a9f)

I had to modify docker-compose.yml to point redis back to the dockerfile. I tried running the dockerfile(s) under tests, but that didn't work at all. `dockerfile: ../../.devcontainer/Dockerfile`
![image](https://github.com/user-attachments/assets/07256dcb-6e67-45bc-98df-fa73640cda36)

I added dotnet 8 installation to the Dockerfile under .devcontainer
RUN curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --channel 8.0 --install-dir /usr/share/dotnet/

kijanawoodard commented:
I wasn't able to get dev containers running. The dockerfile in .devcontainer refers to a Dockerfile in `tests/RedisConfigs` that doesn't exist. I tried using the other ones, but they appear to depend on the `docker-compose.yml` file in the `tests/RedisConfigs` folder.

I was able to run the tests with only one failure: `FailoverTests.SubscriptionsSurviveConnectionFailureAsync(RESP2) [FAIL]`

I got that far by cloning the project in wsl2, running `docker compose up` in `tests/RedisConfigs` and then running `dotnet test StackExchange.Redis.sln`  from the project root.

kijanawoodard commented:
I was wrong about appveyor having no test failures. There were 3 failures that don't seem to be related to my changes here. If there are known test issues, let me know.

kijanawoodard commented:
@NickCraver @mgravell sorry to @ you guys, but I want to make sure I haven't missed some step in the PR process. I looked for a contributing doc, but didn't see one. Is there anything else I need to do at the moment, or is it fine to just let this PR sit tight?

WeihanLi commented:
There's a failed test case, but think it's not related to the changes
![image](https://github.com/user-attachments/assets/a81b1847-bd10-4d43-9b2b-5c41b95f64ac)

Maybe add an entry for the release note

https://github.com/StackExchange/StackExchange.Redis/blob/main/docs/ReleaseNotes.md

mgravell commented:
I very much agree with you about the docs being less than clear on this detail.

kijanawoodard commented:
The original comment for the MINID PR also brings in LIMIT to work with ~. 
Seems definitive to me as it appears, from reading the entire thread, that the top comment was updated as they made decisions on the PR.

https://github.com/redis/redis/pull/8169

kijanawoodard commented:
> the API shape gets tricky, because there is an inherent ambiguity since RedisValue is convertable from int; I wonder whether an explicit API name might be preferable, for example StreamTrimByMinId

I agree. An explicit API makes sense.

kijanawoodard commented:
Side note: more XTRIM options were added a couple days ago:
https://github.com/redis/redis/pull/14130

kijanawoodard commented:
@mgravell 

I changed the method name to StreamTrimByMinId and implemented useApproximateMaxLength and limit.
From the docs, limit is only in play when useApproximateMaxLength = true. (Noted in method documentation)
I made limit nullable because Zero means "no limits". (Noted in method documentation)

The test with useApproximateMaxLength sets the limit to 100. Less than 100, deleted zero items.
I didn't do a test without a limit since I think that's non-deterministic.

Hmm. Writing this up here, useApproximateMaxLength is tied to the MaxLength version of XTRim. Maybe I should rename?

Additional questions:
0. Rename useApproximateMaxLength for the MINID overload?
1. The original purpose of LIMIT was to action on the MAXLENGTH version as well. Implement that here or do that in another PR?
2. Should we throw if the limit is set without setting useApproximateMaxLength since it will have no effect or should we let the documentation suffice?

mgravell commented:
Looking at this now; there are some additional things proposed in 8.2 that I'm looking at
